### PR TITLE
[Fix-15828] fix expandListParameter

### LINKED ISSUE #15828: [Bug] Parameter replace with ? will cause NPE
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

When I want to pass parameters and "?" appears inside the SQL statement, the scheduler will report a null pointer error to me.

### What you expected to happen

I think that when passing parameters, during the escaping process, the parameters will be replaced with "?" first. At this time, it conflicts with the "?" inside the SQL statement.

### How to reproduce

There is no such problem in older versions

### Anything else

my demo like this
select concat('(?i)', (select name from namelist where tag = '${tag}'))

### Version

3.2.x

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xxzuo**
## Purpose of the pull request

- fix #15828 
- Expand the list parameters when matching `${}` instead of first replacing ${} with a `?` and then processing the parameters based on the `?`

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java [hunk 1]
```diff
@@ -454,9 +454,9 @@ private SqlBinds getSqlAndSqlParamsMap(String sql) {
         String rgexo = "['\"]*\\!\\{(.*?)\\}['\"]*";
         sql = replaceOriginalValue(sql, rgexo, paramsMap);
         // replace the ${} of the SQL statement with the Placeholder
-        String formatSql = sql.replaceAll(rgex, "?");
+        // String formatSql = sql.replaceAll(rgex, "?");
         // Convert the list parameter
-        formatSql = ParameterUtils.expandListParameter(sqlParamsMap, formatSql);
+        String formatSql = ParameterUtils.expandListParameter(sqlParamsMap, sql, rgex);
```
Comment (ruanwenjun) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java]: The regex shouldn't exposed to outside.

Comment (xxzuo) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java]: I have reused `AbstractTask.rgex` here. Should I reference `AbstractTask.rgex` in `ParameterUtils` or should I extract this rgex into a constant class? I apologize for not being familiar with this code specification.

Comment (xxzuo) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java]: I extract this rgex into `TaskConstants`

**Review by SbloodyS — APPROVED** [review]
LGTM

**Review by ruanwenjun — APPROVED** [review]

**Pull request merged**